### ISSUE TITLE
DOC: Additional items for the cheat sheet

### ISSUE THREAD

**Issue by Dr-Irv** [pandas-dev/pandas#40680]
#### Location of the documentation

https://pandas.pydata.org/Pandas_Cheat_Sheet.pdf

#### Potential Cheat Sheet Improvements

Per discussion at https://github.com/pandas-dev/pandas/pull/39806#pullrequestreview-590105681 , add a third page to the cheat sheet:

- More visualization examples that use pandas plotting (no dependence on third party libraries)
- List of frequently used options shown here: https://pandas.pydata.org/pandas-docs/stable/user_guide/options.html#frequently-used-options
- I/O: A whole section showing a variety of popular IO usage (CSV, Excel, SQL, HTML), and also output file formats (feather, parquet, HDF)
- An Apply Functions section 
- The new Extension Types (String, Integer, Float) and how pd.NA works
- Anything else that could fill up the space (if needed)

**Comment by kunal21sinha** [comment]
I am first time contributor and would like to take up this issue. 
Could you assign this to me.

**Comment by Dr-Irv** [comment]
@kunal21sinha I've assigned it to you but I know that @OliEfr is possibly working on this based on the discussion from the PR he created.

**Comment by OliEfr** [comment]
Hello,
yes, I was working on the cheatsheet before. However, I did not start working on this precise issue. I think you can go ahead and create something according to your thoughts @kunal21sinha. We will see what additions I can make.

**Comment by kunal21sinha** [comment]
Okay sure, will work on it.

**Comment by AlexGCas** [comment]
maybe you want to add other functions of loc and iloc, iloc for reassignment of a row and loc  for reassingment and append new row

**Comment by amaru-g** [comment]
Hi, This is the very first time I am contributing. according to wiki. I should figure it out if this is is taken or not. @AlexGCas . What is the status of this?
Regards,

**Comment by rishitbhojak** [comment]
@Dr-Irv Can you let me know how can I get started with contributing to this cheatsheet. Like, how do I make changes to this cheatsheet ? I have learnt Data Analysis recently and Pandas was of great help.  I can add more plotting functions from a pandas dataframe.

**Comment by OliEfr** [comment]
@rishitbhojak 
Well, I can help you with that. Go to this folder: https://github.com/OliEfr/pandas/tree/master/doc/cheatsheet
Here you will find the powerpoint, which are used to create the pdf files. Its also useful to read the README.txt.
Besides that of course, it is usefull to know how git and github works.

**Comment by rishitbhojak** [comment]
@Dr-Irv @OliEfr Can I start with pie chart plots and subplots . It seems to be a good idea because we'll add more visualization examples. What's your opinion? Can I add the third page and move forward with it?

**Comment by Dr-Irv** [comment]
@rishitbhojak Add a third page with respect to plotting.  Move the examples at the bottom of page 2 to page 3, but find something to fill the empty space on page 2 that will be created by the move of those 2 examples.  Probably a list of the frequently used options would fit nicely there.

Follow the instructions for contributing to pandas here;  https://pandas.pydata.org/docs/development/contributing.html

**Comment by rishitbhojak** [comment]
Alright sir. Can you let me know whether I have done it correctly or not? I am attaching a sample screenshot
![sample](https://user-images.githubusercontent.com/51954932/123106688-7929f180-d456-11eb-959f-07e8f2b3031e.PNG)

**Comment by Dr-Irv** [comment]
@rishitbhojak that's fine.  Try to keep the graphics smaller.  For the `subplots` example, ideally there would be titles on each subplot, so you should show how to do that.

**Comment by rishitbhojak** [comment]
![image](https://user-images.githubusercontent.com/51954932/123510943-f4301980-d69b-11eb-9ca2-3fb529cfcb1f.png)
Took the visualization portion on the third page and will put some function which we use regularly in place of it. Also, I added titles to both the subplots of the pie chart

**Comment by Dr-Irv** [comment]
@rishitbhojak Any review will occur in a pull request, so when you are all done with your changes, create the PR, and I'll provide more feedback there.

**Comment by rishitbhojak** [comment]
In which branch should I make the pull request? I am done with the plotting pie charts portion and in place of the scatter plot and the histogram, I made a section to drop a dataframe column

**Comment by Dr-Irv** [comment]
> In which branch should I make the pull request? I am done with the plotting pie charts portion and in place of the scatter plot and the histogram, I made a section to drop a dataframe column. I will make the PR in this week

Create your own branch off of master.  See https://pandas.pydata.org/docs/development/contributing.html#working-with-the-code

**Comment by Nishita5421** [comment]
Is this close ? or Can I contribute in this.?

**Comment by Dr-Irv** [comment]
> Is this close ? or Can I contribute in this.?

There is a PR #43036 that was just submitted that I need to review.  Once that is done, I will update the list above with the open items.

**Comment by KeeratKG** [comment]
@Dr-Irv Hello. I'd like to do my bit here. May I take up this issue?

**Comment by KeeratKG** [comment]
> > @Dr-Irv Hello. I'd like to do my bit here. May I take up this issue?
> 
> Thanks for offering your help. I just realized from your note that I had forgotten to review #43036 . So let me get that done and merged, and then you could do additional things from there.

Absolutely! I kind of realised that #43036 might still be in the queue, so I went through its contents and made sure I wasn't repeating anything :)

**Comment by KeeratKG** [comment]
Hi. I have added a fourth page to the cheatsheet after #43036. Please have a look @Dr-Irv and let me know what you think. Will update the contribution accordingly. Refer #45347. 
I wasn't sure what 'The new Extension Types (String, Integer, Float) and how pd.NA works' referred to. Tried to cover everything else.

**Comment by MichaelTiemannOSC** [comment]
Some additional suggestions to consider for this round or the next:

- [ ] wide_to_long: more general and more powerful than melt and important to know it exists
- [ ] groupby box should at least mention the concept of split-apply-combine
- [ ] groupby.first() operator: important for dealing with missing data
- [ ] Group by: split-apply-combine (which brings in the the whole concept of operations on data).  The elaboration of these concepts will easily fill half to a full page with high-quality content.
- [ ] The confusing topic of using conditionals in pandas (what to do with a.any, a.all, etc and when they can be avoided entirely)

**Comment by geo7** [comment]
@MichaelTiemannOSC could you explain how `wide_to_long` is "more general" and "more powerful" than `melt`? As I thought the opposite was true, and the documentation (https://pandas.pydata.org/docs/reference/api/pandas.wide_to_long.html) seems to support that "_ Less flexible but more user-friendly than melt_". Maybe there's something I'm unaware of though.

**Comment by MichaelTiemannOSC** [comment]
Good question.  melt is "more general" in that it can handle multilevel indexes.  wide_to_long is more user-friendly in that in the single-index case, it can both reshape and rename columns, whereas melt only concerns itself with reshaping data.

**Comment by geo7** [comment]
Hrm - to be honest I've always just used `melt` - but `wide_to_long` does have a couple of things packed in that I might use other methods for, I find melt clearer - perhaps because that's what i typically use though. Sorry to derail the thread - I'll leave an example from the `wide_to_long` documentation with a `melt` version incase that's of use to anyone in future.

`data`
```python
df = pd.DataFrame(
    {
        "famid": [1, 1, 1, 2, 2, 2, 3, 3, 3],
        "birth": [1, 2, 3, 1, 2, 3, 1, 2, 3],
        "ht_one": [2.8, 2.9, 2.2, 2, 1.8, 1.9, 2.2, 2.3, 2.1],
        "ht_two": [3.4, 3.8, 2.9, 3.2, 2.8, 2.4, 3.3, 3.4, 2.9],
    }
)
```

`wide_to_long`
```python
pd.wide_to_long(
    df,
    stubnames="ht",
    i=["famid", "birth"],
    j="age",
    sep="_",
    suffix=r"\w+",
)
```

`melt`
```python
(
    df.melt(id_vars=["famid", "birth"], value_name="ht")
    .assign(age=lambda df: df["variable"].str.split("_").str[-1])
    .drop("variable", axis=1)
    .sort_values(["famid", "birth", "age"])
    .set_index(["famid", "birth", "age"])
)
```

**Comment by MichaelTiemannOSC** [comment]
For the record, I use melt 90% of the time.  But there's lots of financial data of the form xyz_2016, xyz_2017, ... and wide_to_long is great for that.

**Comment by suryakapurothu** [comment]
Is it open now?
@geo7 @Dr-Irv 
A bit of guidance would let me add up value to the cheatsheet getting cooked.

**Comment by ashishbalti4** [comment]
@Dr-Irv I have worked and updated the cheat sheet in pdf format. Can I contribute here as it will be my first open-source contribution.

**Comment by Dr-Irv** [comment]
> @Dr-Irv I have worked and updated the cheat sheet in pdf format. Can I contribute here as it will be my first open-source contribution.

@ashishbalti4 You should create a pull request that includes the updated powerpoint and PDF.

**Comment by abrowne34** [comment]
Has this been completed or can I take?

**Comment by Aryabhatt1234** [comment]
Hello @Dr-Irv, I am a first time contributor. Can I take up this issue? Please allow me.

**Comment by Dr-Irv** [comment]
> Hello @Dr-Irv, I am a first time contributor. Can I take up this issue? Please allow me.

Yes, just provide your edits in a PR and I will review.  The PR should update the Powerpoint and the PDF.

**Comment by rootsmusic** [comment]
I suggest specifying the minimum version of pandas that's needed.

**Comment by Dr-Irv** [comment]
> I suggest specifying the minimum version of pandas that's needed.

I think, but I'm not sure, that all of the examples in the cheat sheet work with version 1.x on up.  If you see otherwise, let me know

**Comment by samuel-davidson** [comment]
Hi,
I am new to open source software and would love to contribute.  I have added a third page to the cheatsheet and submitted a PR.  Thank you!

**Comment by MichaelTiemannOSC** [comment]
Could you link to the PR...I'd be interested.  The list of suggestions I made much earlier in this thread (https://github.com/pandas-dev/pandas/issues/40680#issuecomment-1013238952) is not being ticked off (though one topic--melt vs. wide_to_long was discussed in detail). split-transform-combine remains a major concept in data science and is nowhere referenced or exemplified in the Cheat Sheet.

**Comment by samuel-davidson** [comment]
Here is the [PR](https://github.com/pandas-dev/pandas/pull/59469)!.  I just started on the third page and added the frequently used options from the initial list.

**Comment by Sarvvesh123** [comment]
Hi,  I would like to contribute to the proposed 3rd page of the pandas cheat sheet
I plan on including, Apply functions (df.apply())
python:
#apply a function to each column 
df.apply(np.mean)
#apply a function to each row 
df.apply(lambda row: row.max() - row.min(). axis=1)

#elementwise apply
df[ ' col ' ].map(str.upper)
df.applymap(lambda x: len(str(x)))

I can submit a PR that adds this block to the cheat sheet content or relevant docs file. Please let me know if you would like it submitted. 
Thanks!

**Comment by Dr-Irv** [comment]
> I can submit a PR that adds this block to the cheat sheet content or relevant docs file. Please let me know if you would like it submitted.

#61378 may already be addressing this, and is almost complete.  Once that is done, see if those changes haven't addressed your suggestions.
